File: backend/src/main/java/com/imedba/modules/diploma/entity/Diploma.java
package com.imedba.modules.diploma.entity;

import com.imedba.common.entity.BaseEntity;
import com.imedba.modules.course.entity.Course;
import com.imedba.modules.staff.entity.Staff;
import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.FetchType;
import jakarta.persistence.JoinColumn;
import jakarta.persistence.JoinTable;
import jakarta.persistence.ManyToMany;
import jakarta.persistence.OneToMany;
import jakarta.persistence.OrderBy;
import jakarta.persistence.Table;
import java.util.ArrayList;
import java.util.List;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Builder.Default;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;
import org.hibernate.annotations.SQLDelete;
import org.hibernate.annotations.SQLRestriction;

@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
@Builder
@Entity
@Table(name = "diplomas")
@SQLDelete(sql = "UPDATE diplomas SET deleted_at = NOW() WHERE id = ?")
@SQLRestriction("deleted_at IS NULL")
public class Diploma extends BaseEntity {

    @Column(name = "name", nullable = false, length = 300)
    private String name;

    @Column(name = "university_name", length = 200)
    private String universityName;

    @Column(name = "description", columnDefinition = "TEXT")
    private String description;

    /** Comisiones (V044): cursos FS a los que se inscriben los alumnos; la liquidación suma los pagos de todas. */
    @Default
    @OneToMany(mappedBy = "diploma", fetch = FetchType.LAZY)
    @OrderBy("commission DESC")
    private List<Course> commissions = new ArrayList<>();

    /**
     * Directoras de la diplomatura, tomadas de Personal Académico (V035).
     *
     * <p><b>Sin porcentaje.</b> Antes se pedía un «% de directora» al crear la
     * diplomatura; el cliente lo bajó explícitamente el 2026-07-23 («eso habría que
     * sacarlo y que sólo pida cuántas directoras y quiénes»). Se reparten en partes
     * iguales la mitad del subtotal 2 menos las grabaciones.
     *
     * <p>Todos los costos y porcentajes de la liquidación viven en el settlement,
     * no acá (decisión 2026-05-22 §2.6).
     */
    @Default
    @ManyToMany(fetch = FetchType.LAZY)
    @JoinTable(
            name = "diploma_directors",
            joinColumns = @JoinColumn(name = "diploma_id"),
            inverseJoinColumns = @JoinColumn(name = "staff_id"))
    private List<Staff> directors = new ArrayList<>();

    @Default
    @Column(name = "is_active", nullable = false)
    private Boolean active = Boolean.TRUE;
}

File: frontend/src/types/diploma.ts
import type { Instant, UUID } from './common'
import type { Modality } from './course'

/**
 * Directora de la diplomatura. Viene de Personal Académico (staff con rol
 * DIRECTORA). **Sin porcentaje**: se reparten en partes iguales la mitad del
 * subtotal 2 menos las grabaciones docentes.
 *
 * Antes se pedía un «% de la directora» al crear la diplomatura; el cliente lo
 * bajó el 2026-07-23.
 */
export interface DirectorRef {
  id:    UUID
  name:  string
  email: string | null
}

// Refleja CommissionResponse: `id` es el del curso FS al que se inscriben los alumnos.
export interface Commission {
  id:                UUID
  name:              string
  commission:        number | null  // null = curso espejo viejo (pre-V044) sin número todavía
  academicYear:      number | null
  enrollmentPrice:   number | null
  coursePrice:       number | null
  includesPremaBook: boolean | null
  startDate:         string | null   // ISO yyyy-mm-dd
  endDate:           string | null
  modality:          Modality | null
  moodleCourseId:    number | null
  active:            boolean | null
}

// Refleja CommissionRequest
export interface CommissionRequest {
  commission:         number
  academicYear?:      number | null
  enrollmentPrice?:   number | null
  coursePrice?:       number | null
  includesPremaBook?: boolean | null
  startDate?:         string | null
  endDate?:           string | null
  modality?:          Modality | null
  moodleCourseId?:    number | null
  active?:            boolean | null
}

// Refleja DiplomaResponse: la diplomatura es el programa general; precios, libro y fechas van por comisión.
export interface Diploma {
  id:             UUID
  name:           string
  universityName: string | null
  description:    string | null
  directors:      DirectorRef[]
  commissions:    Commission[]    // de la más nueva a la más vieja
  active:         boolean | null
  createdAt:      Instant
  updatedAt:      Instant
}

// Refleja DiplomaCreateRequest
export interface DiplomaCreateRequest {
  name:             string          // required, max 300
  universityName?:  string | null   // max 200
  description?:     string | null
  /** Ids de Personal Académico con rol DIRECTORA. */
  directorIds?:     UUID[] | null
  firstCommission?: CommissionRequest | null
}

// Refleja DiplomaUpdateRequest. `directorIds` en null = no tocar; lista (incluso vacía) = reemplaza el set.
export interface DiplomaUpdateRequest {
  name?:           string | null
  universityName?: string | null
  description?:    string | null
  directorIds?:    UUID[] | null
  active?:         boolean | null
}
